infrastructure/src/main/java/org/corfudb/infrastructure/management/HealingDetector.java

```diff
@@ -0,0 +1,233 @@
+        }
+    }
+
+    /**
+     * Reset all responses to an invalid iteration number, -1.
+     */
+    private void resetResponses() {
+        for (int i = 0; i < responses.length; i++) {
+            responses[i] = INVALID_RESPONSE;
+        }
+    }
+
+    /**
+     * Block on all poll futures and collect the responses. There are 3 possible cases.
+     * 1. Receive a PONG. We set the responses[i] to the polling iteration number.
+     * 2. WrongEpochException is thrown. We mark as a successful response.
+     * 3. Other Exception is thrown. We do nothing. (The response[i] in this case is left behind.)
+     *
+     * @param pollIteration poll iteration in the ongoing polling round.
+     */
+    private void collectResponsesAndVerifyEpochs(int pollIteration) {
+        // Collect responses and increment response counters for successful pings.
+        for (int i = 0; i < members.length; i++) {
+            try {
+                pollCompletableFutures[i].get();
+                responses[i] = pollIteration;
+            } catch (Exception e) {
+                // WrongEpochException signifies liveness of node.
+                if (e.getCause() instanceof WrongEpochException) {
+                    responses[i] = pollIteration;
+                }
+            }
+        }
+    }
+
+    /**
+     * Aggregates all responses and checks if any node healed to rejoin the cluster.
+     *
+     * @param membersSet Set of unresponsive member servers
+     * @return True if nodes to heal present. Else False.
+     */
+    private boolean isHealedNodesPresent(Set<String> membersSet) {
+        // Count unsuccessful ping responses.
+        Set<String> unresponsiveNodes = new HashSet<>();
+        for (int i = 0; i < members.length; i++) {
+            if (responses[i] == INVALID_RESPONSE) {
+                unresponsiveNodes.add(members[i]);
+            }
+        }
+
+        // If we receive all responses and there are no nodes to heal,
+        // we can assume there are no new faults and exit.
+        return !unresponsiveNodes.equals(membersSet);
+    }
+
+    /**
+     * Each PollRound executes "detectionThreshold" number of iterations and generates the
+     * poll report.
+     *
+     * @return Poll Report with detected healed nodes.
+     */
+    private PollReport pollRound() {
+
+        Set<String> membersSet = new HashSet<>(Arrays.asList(members));
+        Set<String> healed = new HashSet<>();
+        resetResponses();
+
+        // In each iteration we poll all the servers in the members list.
+        for (int iteration = 1; iteration <= detectionThreshold; iteration++) {
+
+            // Ping all nodes and await their responses.
+            pollOnceAsync();
+
+            // Collect all poll responses.
+            collectResponsesAndVerifyEpochs(iteration);
+
+            if (!isHealedNodesPresent(membersSet)) {
+                break;
+            }
+
+            try {
+                Thread.sleep(interIterationInterval);
```

Comment:
why do you have to do this and not use a scheduled executor feature to start after certain delay the next task.

Reply:
This sleep and loop is executed only if there is a failure.
In an ideal scenario, each round consists of only one iteration as there are no failures.
If there are failures, the round confirms this failure by iterating thrice.
So, from my view, this reduces a bit of code complexity. Let me know if you feel otherwise.